Mage/src/main/java/mage/Mana.java

```diff
@@ -1323,9 +1372,27 @@ public boolean equals(Object o) {
                 && any == mana.any;
     }
 
+    /**
+     * Hardcoding here versus using Objects.hash in order to increase performance since this is
+     * called thousands of times by {@link mage.abilities.mana.ManaOptions#addManaWithCost(List, Game)}
+     *
+     * @return
+     */
     @Override
     public int hashCode() {
-        return Objects.hash(white, blue, black, red, green, generic, colorless, any, flag);
+        int result = 1;
+
+        result = 31 * result + white;
+        result = 31 * result + blue;
+        result = 31 * result + black;
+        result = 31 * result + red;
+        result = 31 * result + green;
+        result = 31 * result + generic;
+        result = 31 * result + colorless;
+        result = 31 * result + any;
+        result = 31 * result + (flag ? 1 : 0);
+
+        return result;
```

Comment:
Hashing for Mana is going to overflow for large color values. Each mana color is allowed to go as high as `Integer.MAX_VALUE`.

Perhaps something from: https://stackoverflow.com/questions/4045063/how-should-i-map-long-to-int-in-hashcode